CLI: Migrate CLI to TypeScript

gaetanmaisse opened the pull request:
Somehow related to https://github.com/storybooks/storybook/issues/5030

## What I did

Migrate `@storybook/cli` to TypeScript.

I tried to introduce as little change as possible because we are already in the beta phase. The goal of this PR is just to migrate the code to TS to simplify future refactoring and work on E2E and CLI (see https://github.com/storybookjs/storybook/issues/10723 for details). Some refactoring will come later. 

Sidenote: there are a lot of commits because I worked with small steps, if needed I can squash them in a single one.

## How to test

 - All CLI tests must be 🟢

gaetanmaisse committed:
refactor(cli): add `tsconfig.json`

⚠️ All files in `template*` directories are ignored because they must be as it in the `dist`.

gaetanmaisse committed:
refactor(cli): migrate Yarn and NPM related files to TS

 - Add needed @types packages
 - remove `silent: true` option in `sync` calls as it looks like it's not a valid option
 - remove `encoding: 'utf-8'` option in `spawn` calls as it looks like it's not a valid option

gaetanmaisse committed:
refactor(cli): migrate project_types and its tests to TS

gaetanmaisse committed:
refactor(cli): migrate helpers and its tests to TS

gaetanmaisse committed:
refactor(cli): migrate `add` file and its tests to TS

gaetanmaisse committed:
refactor(cli): migrate `detect` file and its tests to TS

gaetanmaisse committed:
refactor(cli): migrate `extract` file to TS

gaetanmaisse committed:
refactor(cli): migrate `migrate` file to TS

gaetanmaisse committed:
refactor(cli): move NpmOptions and PackageJson in their own files

gaetanmaisse committed:
refactor(cli): migrate Angular generator to TS

gaetanmaisse committed:
refactor(cli): migrate Aurelia generator to TS

gaetanmaisse committed:
refactor(cli): migrate Ember generator to TS

gaetanmaisse committed:
refactor(cli): migrate HTML generator to TS

gaetanmaisse committed:
refactor(cli): migrate Marionette generator to TS

gaetanmaisse committed:
refactor(cli): migrate Marko generator to TS

gaetanmaisse committed:
refactor(cli): migrate Meteor generator to TS

gaetanmaisse committed:
refactor(cli): migrate Mithril generator to TS

gaetanmaisse committed:
refactor(cli): migrate Preact generator to TS

gaetanmaisse committed:
refactor(cli): migrate Rax generator to TS

gaetanmaisse committed:
refactor(cli): migrate React generator to TS

gaetanmaisse committed:
refactor(cli): migrate React Native generator to TS

gaetanmaisse committed:
refactor(cli): migrate React Scripts generator to TS

gaetanmaisse committed:
refactor(cli): migrate Riot generator to TS

gaetanmaisse committed:
refactor(cli): migrate Sfc Vue generator to TS

gaetanmaisse committed:
refactor(cli): migrate Svelte generator to TS

gaetanmaisse committed:
refactor(cli): migrate Vue generator to TS

gaetanmaisse committed:
refactor(cli): migrate Web Components generator to TS

gaetanmaisse committed:
refactor(cli): migrate Webpack React generator to TS

gaetanmaisse committed:
refactor(cli): migrate update package organization generator to TS

gaetanmaisse committed:
refactor(cli): migrate `initiate` file to TS

gaetanmaisse committed:
refactor(cli): migrate `generate` file to TS

gaetanmaisse committed:
refactor(cli): activate strict mode in TS (except `strictNullChecks`)

Activating `strictNullChecks` makes TS compilation throw a lot of errors that need to refactor some part of the code.
These refactoring will be done in another commit.

yannbf committed:
refactor(cli): improve typings for project types

yannbf committed:
refactor(cli): improve typings for story format

yannbf committed:
refactor(cli): improve typings for supported language

yannbf commented:
This is golden 💕

shilman reviewed (commented):
Great job @gaetanmaisse 💯

yannbf committed:
refactor(cli): add storyformat to command options

yannbf committed:
test(cli): fix types by casting storyformat

yannbf committed:
refactor(cli): use storyformat in aurelia generator

gaetanmaisse committed:
refactor(cli): use a unified type across all CLI generators

gaetanmaisse committed:
refactor(cli): improve typing of `window`

In order to be able to extend types in global scope with an `import` at the top of the file we have to create a new dedicated file (and can not use already existing typings.d.ts).
For details, see https://stackoverflow.com/questions/30960386/how-to-extend-the-window-typescript-interface/30961346#30961346

gaetanmaisse committed:
Merge branch 'next' into tech/migrate-cli-to-TS

gaetanmaisse commented:
Thanks for the help @yannbf. I just added 2 commits related to the feedbacks given by @tooppaaa and resolved conflicts with `next` so as soon as CI is 🟢, it is mergeable 🎉
